File: components/deno/Constraints.test.ts
import { assertEquals } from "https://deno.land/std@0.137.0/testing/asserts.ts";
import { Constraints } from "./Constraints.ts";
import { inferProgram } from "./Infer.ts";
import { parse } from "./Parser.ts";
import { createFresh, emptyTypeEnv, Type } from "./Typing.ts";

const solve = (expression: string): Array<Type> => {
  const [constraints, types] = inferProgram(
    emptyTypeEnv,
    parse(expression),
    new Constraints(),
    createFresh(),
  );

  const subst = constraints.solve();

  return types.map((t) => t.apply(subst));
};

const assertType = (expression: string, expected: Array<string>) => {
  assertEquals(
    expected.join(","),
    solve(expression).toString(),
  );
};

Deno.test("solve \\x -> \\y -> \\z -> x + y + z", () => {
  assertType(
    "\\x -> \\y -> \\z -> x + y + z",
    ["Int -> Int -> Int -> Int"],
  );
});

Deno.test("solve \\f -> \\g -> \\x -> f (g x)", () => {
  assertType(
    "\\f -> \\g -> \\x -> f (g x)",
    ["(V4 -> V5) -> (V3 -> V4) -> V3 -> V5"],
  );
});

Deno.test("solve match (1, 2) with (x, y) -> x + y", () => {
  assertType(
    "match (1, 2) with (x, y) -> x + y",
    ["Int"],
  );

  assertType(
    "let x = (1, (False, 99)) ; match x with (_, (False, x)) -> x | (x, (True, _)) -> x",
    ["((Int * (Bool * Int)))", "Int"],
  );
});

Deno.test("solve let rec? compose = \\f -> \\g -> \\x -> f (g x) ; compose", () => {
  assertType(
    "let compose = \\f -> \\g -> \\x -> f (g x) ; compose",
    [
      "((V4 -> V5) -> (V3 -> V4) -> V3 -> V5)",
      "(V6 -> V7) -> (V8 -> V6) -> V8 -> V7",
    ],
  );

  assertType(
    "let rec compose = \\f -> \\g -> \\x -> f (g x) ; compose",
    [
      "((V6 -> V7) -> (V5 -> V6) -> V5 -> V7)",
      "(V9 -> V10) -> (V11 -> V9) -> V11 -> V10",
    ],
  );
});

Deno.test("solve let rec? f = (\\x -> x) ; let rec? g = (f True) ; f 3", () => {
  assertType(
    "let f = (\\x -> x) ; let g = (f True) ; f 3",
    ["(V1 -> V1)", "(Bool)", "Int"],
  );

  assertType(
    "let f = (\\x -> x) ; let rec g = (f True) ; f 3",
    ["(V1 -> V1)", "(Bool)", "Int"],
  );

  assertType(
    "let rec f = (\\x -> x) ; let g = (f True) ; f 3",
    ["(V3 -> V3)", "(Bool)", "Int"],
  );

  assertType(
    "let rec f = (\\x -> x) ; let rec g = (f True) ; f 3",
    ["(V3 -> V3)", "(Bool)", "Int"],
  );
});

Deno.test("solve let rec? identity = \\n -> n ; identity", () => {
  assertType(
    "let identity = \\n -> n ; identity",
    ["(V1 -> V1)", "V2 -> V2"],
  );

  assertType(
    "let rec identity = \\n -> n ; identity",
    ["(V3 -> V3)", "V5 -> V5"],
  );
});

Deno.test("solve let rec? add a b = a + b and succ = add 1 ; succ 10", () => {
  assertType(
    "let add a b = a + b and succ = add 1 ; succ 10",
    ["(Int -> Int -> Int * Int -> Int)", "Int"],
  );

  assertType(
    "let rec add a b = a + b and succ = add 1 ; succ 10",
    ["(Int -> Int -> Int * Int -> Int)", "Int"],
  );
});

Deno.test("solve let rec fact n = if (n == 0) 1 else fact (n - 1) * n ; fact", () => {
  assertType(
    "let rec fact n = if (n == 0) 1 else fact(n - 1) * n ; fact",
    ["(Int -> Int)", "Int -> Int"],
  );
});

Deno.test("solve let rec isOdd n = if (n == 0) False else isEven (n - 1) and isEven n = if (n == 0) True else isOdd (n - 1) ; isOdd", () => {
  assertType(
    "let rec isOdd n = if (n == 0) False else isEven (n - 1) and isEven n = if (n == 0) True else isOdd (n - 1) ; isOdd",
    ["(Int -> Bool * Int -> Bool)", "Int -> Bool"],
  );
});

Deno.test("solve let rec a = b + 1 and b = a + 1 ; a", () => {
  assertType(
    "let rec a = b + 1 and b = a + 1 ; a",
    ["(Int * Int)", "Int"],
  );
});

Deno.test("solve let rec? identity a = a and v = identity 10 ; v", () => {
  assertType(
    "let identity a = a and v = identity 10 ; v",
    ["(V1 -> V1 * Int)", "Int"],
  );

  assertType(
    "let rec identity a = a and v = identity 10 ; v",
    ["(Int -> Int * Int)", "Int"],
  );
});

Deno.test("solve let rec? identity a = a ; let rec? v1 = identity 10 and v2 = identity True ; v?", () => {
  assertType(
    "let identity a = a ; let rec v1 = identity 10 and v2 = identity True ; v1",
    ["(V1 -> V1)", "(Int * Bool)", "Int"],
  );

  assertType(
    "let rec identity a = a ; let rec v1 = identity 10 and v2 = identity True ; v1",
    ["(V3 -> V3)", "(Int * Bool)", "Int"],
  );

  assertType(
    "let identity a = a ; let rec v1 = identity 10 and v2 = identity True ; v2",
    ["(V1 -> V1)", "(Int * Bool)", "Bool"],
  );

  assertType(
    "let rec identity a = a ; let rec v1 = identity 10 and v2 = identity True ; v2",
    ["(V3 -> V3)", "(Int * Bool)", "Bool"],
  );
});

File: components/kotlin/app/src/test/kotlin/tlca/ConstraintsTest.kt
package tlca

import kotlin.test.Test
import kotlin.test.assertEquals

class ConstraintsTest {
    @Test
    fun `lambda function with arithmetic operators`() {
        assertType(
            listOf("Int -> Int -> Int -> Int"),
            "\\x -> \\y -> \\z -> x + y + z"
        )
    }

    @Test
    fun `lambda function composition`() {
        assertType(
            listOf("(V4 -> V5) -> (V3 -> V4) -> V3 -> V5"),
            "\\f -> \\g -> \\x -> f (g x)"
        )
    }

    @Test
    fun `returns compose function where compose is defined using lambda`() {
        assertType(
            listOf("((V4 -> V5) -> (V3 -> V4) -> V3 -> V5)", "(V6 -> V7) -> (V8 -> V6) -> V8 -> V7"),
            "let compose = \\f -> \\g -> \\x -> f (g x) ; compose"
        )

        assertType(
            listOf("((V6 -> V7) -> (V5 -> V6) -> V5 -> V7)", "(V9 -> V10) -> (V11 -> V9) -> V11 -> V10"),
            "let rec compose = \\f -> \\g -> \\x -> f (g x) ; compose"
        )
    }

    @Test
    fun `generalised inferred scheme used in different forms`() {
        assertType(
            listOf("(V1 -> V1)", "(Bool)", "Int"),
            "let f = (\\x -> x) ; let g = (f True) ; f 3"
        )

        assertType(
            listOf("(V3 -> V3)", "(Bool)", "Int"),
            "let rec f = (\\x -> x) ; let g = (f True) ; f 3"
        )

        assertType(
            listOf("(V1 -> V1)", "(Bool)", "Int"),
            "let f = (\\x -> x) ; let rec g = (f True) ; f 3"
        )

        assertType(
            listOf("(V3 -> V3)", "(Bool)", "Int"),
            "let rec f = (\\x -> x) ; let rec g = (f True) ; f 3"
        )
    }

    @Test
    fun `identity declaration and returned`() {
        assertType(
            listOf("(V1 -> V1)", "V2 -> V2"),
            "let identity = \\n -> n ; identity"
        )

        assertType(
            listOf("(V3 -> V3)", "V5 -> V5"),
            "let rec identity = \\n -> n ; identity"
        )
    }

    @Test
    fun `identity declared and used and returned`() {
        assertType(
            listOf("(V1 -> V1 * Int)", "Int"),
            "let identity = \\n -> n and v = identity 10 ; v"
        )

        assertType(
            listOf("(Int -> Int * Int)", "Int"),
            "let rec identity = \\n -> n and v = identity 10 ; v"
        )
    }

    @Test
    fun `sequential declarations`() {
        assertType(
            listOf("(V1 -> V1)", "(Int * Bool)", "Int"),
            "let identity = \\n -> n ; let rec v1 = identity 10 and v2 = identity True ; v1"
        )

        assertType(
            listOf("(V3 -> V3)", "(Int * Bool)", "Int"),
            "let rec identity = \\n -> n ; let rec v1 = identity 10 and v2 = identity True ; v1"
        )

        assertType(
            listOf("(V1 -> V1)", "(Int * Bool)", "Bool"),
            "let identity = \\n -> n ; let rec v1 = identity 10 and v2 = identity True ; v2"
        )

        assertType(
            listOf("(V3 -> V3)", "(Int * Bool)", "Bool"),
            "let rec identity = \\n -> n ; let rec v1 = identity 10 and v2 = identity True ; v2"
        )
    }

    @Test
    fun `sequential declarations with partial application`() {
        assertType(
            listOf("(Int -> Int -> Int * Int -> Int)", "Int"),
            "let add a b = a + b and succ = add 1 ; succ 10"
        )
    }

    @Test
    fun `factorial declaration`() {
        assertType(
            listOf("(Int -> Int)", "Int -> Int"),
            "let rec fact n = if (n == 0) 1 else fact (n - 1) * n ; fact"
        )
    }

    @Test
    fun `mutually recursive function declarations`() {
        assertType(
            listOf("(Int -> Bool * Int -> Bool)", "Int -> Bool"),
            "let rec isOdd n = if (n == 0) False else isEven (n - 1) and isEven n = if (n == 0) True else isOdd (n - 1) ; isOdd"
        )
    }

    @Test
    fun `mutually recursive constant value declarations`() {
        assertType(
            listOf("(Int * Int)", "Int"),
            "let rec a = b + 1 and b = a + 1 ; a"
        )
    }

    private fun assertType(expected: List<String>, expression: String) {
        val (constraints, types) = infer(
            emptyTypeEnv,
            parseExpressions(expression),
            Constraints(),
            Pump()
        )

        assertEquals(expected, types.map { it.apply(constraints.solve()).toString() })
    }
}
